docs/adr/0004-runtime-governance.md

```diff
@@ -0,0 +1,292 @@
+)
+
+// ... some text serialization methods omitted ...
+```
+
+The `governance_model` field can specifiy one of the following governance
+models:
+
+- **Entity governance (`GovernanceEntity`).** This causes the runtime to behave
+  exactly as before, the runtime owner (indicated by `entity_id` in the runtime
+  descriptor) is the only one who can update the runtime descriptor via
+  `registry.RegisterRuntime` method calls.
+
+  The runtime owner is also the one that needs to provide the required stake
+  in escrow in order to avoid the runtime from being suspended. As before note
+  that anyone can delegate the required stake to the runtime owner in order to
+  enable runtime operation (but the owner can always prevent the runtime from
+  operating by performing actions which would cause the stake claims to no
+  longer be satisfied).
+
+- **Runtime-defined governance (`GovernanceRuntime`).** In this case the runtime
+  itself is the only one who can update the runtime descriptor by emitting a
+  runtime message. The runtime owner (indicated by `entity_id`) is not able to
+  perform any updates after the initial registration and such attempts must
+  return `ErrForbidden`.
+
+  The runtime itself is the one that needs to provide the required stake in
+  escrow in order to avoid the runtime from being suspended. This assumes that
+  runtimes can have accounts in the staking module as specified by [ADR 0003].
+  Note that anyone can delegate the required stake to a runtime in order to
+  enable its operation.
+
+- **Consensus layer governance (`GovernanceConsensus`).** In this case only the
+  consensus layer itself can update the runtime descriptor either through a
+  network upgrade or via a consensus layer governance mechanism not specified by
+  this proposal.
+
+  Runtimes using this governance model are never suspended and do not need to
+  provide stake in escrow.
+
+  Runtimes using this governance model cannot be registered/updated via regular
+  registry method calls or runtime messages (doing so must return
+  `ErrForbidden`). Instead such a runtime can only be registered at genesis,
+  through a network upgrade or via a consensus layer governance mechanism not
+  specified by this proposal.
+
+#### Entity Whitelist Admission Policy
+
+The entity whitelist admission policy configuration structure is changed to
+allow specifying the maximum number of nodes that each entity can register under
+the given runtime for each role.
+
+```golang
+type EntityWhitelistConfig struct {
+    // MaxNodes is the maximum number of nodes that an entity can register under
+    // the given runtime for a specific role. If the map is empty or absent, the
+    // number of nodes is unlimited. If the map is present and non-empty, the
+    // the number of nodes is restricted to the specified maximum (where zero
+    // means no nodes allowed), any missing roles imply zero nodes.
+    MaxNodes map[node.RolesMask]uint16 `json:"max_nodes,omitempty"`
```

Comment:
This should probably only allow `RolesMask` values that present a single role (e.g. where single bit is set)? Otherwise there could be conflicting entries.

Reply:
Ah addressed bellow :grin:

Reply:
Yeah this is already specified below:
> Each key (roles mask) in the `max_nodes` map must specify a single role, otherwise the runtime descriptor is rejected with `ErrInvalidArgument`.